### PR TITLE
#256 add discussion review endpoints

### PR DESCRIPTION
**Pull request by Ghaith-Chrit**
This PR closes: #256

This PR adds the following:
- _HTTP_ endpoints to post and view all discussion reviews
- Simplify the DossierDetailsDTO by removing cyclic information
- Add test cases

In the future, _if needed_, WebSockets can be used to provide real-time communications.

Note 1: Apply the new migrations before testing.
Note 2: Any published dossier should be deleted before testing this functionality. This is needed since a "discussion board" is only created once a dossier is published.

### CODE REVIEW

File: Server/ConcordiaCurriculumManager/Models/Curriculum/Dossiers/DossierReview/DiscussionMessage.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+    public required Guid GroupId { get; set; }
+
+    public Group? Group { get; set; }
+
+    public required Guid AuthorId { get; set; }
+
+    public User? Author { get; set; }
+
+    public Guid? ParentDiscussionMessageId { get; set; }
```
Comment (JamesPartsafas) [inline Server/ConcordiaCurriculumManager/Models/Curriculum/Dossiers/DossierReview/DiscussionMessage.cs:21]: Is ParentDiscussionMessage for functionality to reply to messages?

File: Server/ConcordiaCurriculumManager/Repositories/DossierReviewRepository.cs [hunk 2]
```diff
@@ -50,4 +51,20 @@ public async Task<bool> SaveApprovalStages(IList<ApprovalStage> stages)
             .ThenInclude(deletionRequests => deletionRequests.Course)
             .FirstOrDefaultAsync();
     }
+
+    public async Task<Dossier?> GetDossierWithApprovalStagesAndRequestsAndDiscussion(Guid dossierId)
+    {
```
Comment (JamesPartsafas) [inline Server/ConcordiaCurriculumManager/Repositories/DossierReviewRepository.cs:56]: On the frontend, we'll be able to see the ID of the person who sent a message, but how can we retrieve their name or email to show it? I think the easiest option is to remove this method altogether and use the already existing `GetDossierByDossierId` in DossierRepository, but add the extra lines about including the discussion and messages. This way we'll include the group members and the frontend can extract their names from there.

File: Server/ConcordiaCurriculumManager/Services/DossierReviewService.cs [hunk 3]
```diff
@@ -117,10 +130,48 @@ private async Task AcceptDossierChanges(Dossier dossier)
...
+        {
+            throw new BadRequestException("The dossier is not published yet.");
+        }
+
+        return dossier;
+    }
+
+    public async Task AddDossierDiscussionReview(Guid dossierId, DiscussionMessage message)
+    {
+        var userId = _httpContextAccessor.HttpContext?.User.Claims?.FirstOrDefault(c => c.Type.Equals(Claims.Id)) ?? throw new BadRequestException("User Id does not exist");
```
Comment (JamesPartsafas) [inline Server/ConcordiaCurriculumManager/Services/DossierReviewService.cs:147]: There's a method `GetCurrentUserClaim` in UserAuthenticationService to get this, so we don't duplicate code, and then we can remove the reference to httpContextAccessor from the constructor.

File: Server/ConcordiaCurriculumManager/Services/DossierReviewService.cs [hunk 4]
```diff
@@ -11,8 +13,11 @@ public interface IDossierReviewService
     public Task RejectDossier(Guid dossierId);
     public Task ReturnDossier(Guid dossierId);
     public Task ForwardDossier(Guid dossierId);
+    public Task<Dossier> GetDossierWithDiscussion(Guid dossierId);
     public Task<Dossier> GetDossierWithApprovalStagesOrThrow(Guid dossierId);
     public Task<Dossier> GetDossierWithApprovalStagesAndRequestsOrThrow(Guid dossierId);
+    public Task<Dossier> GetDossierWithApprovalStagesAndRequestsAndDiscussionOrThrow(Guid dossierId);
+    Task AddDossierDiscussionReview(Guid dossierId, DiscussionMessage message);
```
Comment (JamesPartsafas) [inline Server/ConcordiaCurriculumManager/Services/DossierReviewService.cs:20]: Missing `public`

**Comment by JamesPartsafas** [comment]
Note additionally that when Zi Hao merges #257, you may get some merge conflicts, or unexpected weirdness, so make sure to merge from main when that's merged into it.

File: Server/ConcordiaCurriculumManager/Services/DossierReviewService.cs [hunk 5]
```diff
@@ -46,6 +54,11 @@ public async Task SubmitDossierForReview(DossierSubmissionDTO dto)
         if (!(await _groupService.IsGroupIdListValid(dto.GroupIds))) throw new InvalidInputException("All groups set as reviewers must be valid groups");
 
         Dossier dossier = await _dossierService.GetDossierDetailsByIdOrThrow(dto.DossierId);
+        dossier.Discussion = new()
```
Comment (JamesPartsafas) [inline Server/ConcordiaCurriculumManager/Services/DossierReviewService.cs:57]: Can we just create the discussion board with the initial dossier creation, that way we don't have to worry about nulls?

File: Server/ConcordiaCurriculumManager/DTO/Dossiers/DossierDetailsWithDiscussionDTO.cs [hunk 6]
```diff
@@ -0,0 +1,8 @@
+﻿using ConcordiaCurriculumManager.DTO.Dossiers.DossierReview;
+
+namespace ConcordiaCurriculumManager.DTO.Dossiers;
+
+public class DossierDetailsWithDiscussionDTO : DossierDetailsDTO
+{
+    public required DossierDiscussionDTO Discussion { get; set; }
```
Comment (JamesPartsafas) [inline Server/ConcordiaCurriculumManager/DTO/Dossiers/DossierDetailsWithDiscussionDTO.cs:7]: If we create the discussion board on initial dossier creation, then this Discussion field can just be part of the DossierDetailsDTO and this file can be removed for simplicity.

File: Server/ConcordiaCurriculumManager/Controllers/DossierReviewController.cs [hunk 7]
```diff
@@ -71,4 +76,28 @@ public async Task<ActionResult> ForwardDossier([FromRoute, Required] Guid dossie
...
+        dossierMessage.DossierDiscussionId = dossierId;
+        await _dossierReviewService.AddDossierDiscussionReview(dossierId, dossierMessage);
+        return NoContent();
+    }
+
+    [HttpPut(nameof(GetAllDossierReviews) + "/{dossierId}")]
+    [SwaggerResponse(StatusCodes.Status422UnprocessableEntity, "Invalid input")]
+    [SwaggerResponse(StatusCodes.Status500InternalServerError, "Unexpected error")]
+    [SwaggerResponse(StatusCodes.Status200OK, "All Dossier Review were successfully returned")]
+    public async Task<ActionResult> GetAllDossierReviews([FromRoute, Required] Guid dossierId)
```
Comment (JamesPartsafas) [inline Server/ConcordiaCurriculumManager/Controllers/DossierReviewController.cs:97]: If we create the discussion board when creating the dossier, then remove the DossierDetailsWithDiscussionDTO as I wrote in other comments, then I think we can just call the service layer method to return the dossier details, and the new service layer method `GetDossierWithDiscussion` is not needed. Likewise, this endpoint is not needed either; the user can just query the dossier details.

File: Server/ConcordiaCurriculumManager/Models/Curriculum/Dossiers/DossierReview/DiscussionMessage.cs (hunk 1 above)
Comment (Ghaith-Chrit) [reply Server/ConcordiaCurriculumManager/Models/Curriculum/Dossiers/DossierReview/DiscussionMessage.cs:21]: Yes

File: Server/ConcordiaCurriculumManager/Controllers/DossierReviewController.cs (hunk 7 above)
Comment (Ghaith-Chrit) [reply Server/ConcordiaCurriculumManager/Controllers/DossierReviewController.cs]: The only reason I created a new endpoint is to reduce the response time. Currently, it takes around 10-20 seconds to retrieve the dossier with discussions (even when I don't have many messages) due to multiple joins. But I'll change it if it's easier for the frontend to just hit a single endpoint.

Comment (JamesPartsafas) [reply Server/ConcordiaCurriculumManager/Controllers/DossierReviewController.cs]: I tested with a dossier with 2 requests, 6 messages, and 2 reviewing groups, and querying the dossier was basically instant on my machine, so I suspect there's some issue on your end (too many programs open simultaneously?)

**Review by JamesPartsafas — APPROVED** [review]
Nice work, seems to work well

**Pull request merged**